File: java/demo/src/test/java/com/yuanzhibang/demo/api/OauthApiHelperTest.java
package com.yuanzhibang.demo.api;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import org.junit.Test;

import com.google.gson.internal.LinkedTreeMap;
import com.yuanzhibang.demo.EnvHelper;

/**
 * Unit test for simple App.
 */
public class OauthApiHelperTest {

    @Test
    public void test_getApiByPath() {
        String api = OauthApiHelper.getApiByPath("/x");
        String expectApi = "https://oauth.yuanzhibang.com/x";
        assertEquals(api, expectApi);
    }

    @Test
    public void test_checkCode_valid() {
        String appId = EnvHelper.getEnvValue("TEST_APP_ID");
        String code = EnvHelper.getEnvValue("TEST_CODE");
        try {
            Object response = OauthApiHelper.checkCode(appId, code, "access_token", null);
            assertEquals(response.getClass(), LinkedTreeMap.class);
        } catch (Exception e) {
            assertTrue(false);
        }
    }

    @Test
    public void test_checkCode_invalid() {
        String appId = EnvHelper.getEnvValue("TEST_APP_ID");
        String code = "123";
        try {
            boolean response = (boolean) OauthApiHelper.checkCode(appId, code, "access_token", null);
            assertFalse(response);
        } catch (Exception e) {
            assertTrue(false);
        }
    }

    @Test
    public void test_getAppUserCount() {
        String appId = EnvHelper.getEnvValue("TEST_APP_ID");
        String code = EnvHelper.getEnvValue("TEST_CODE");
        try {
            int count = OauthApiHelper.getAppUserCount(appId, code, null);
            assertTrue(count >= 0);
        } catch (Exception e) {
            assertTrue(false);
        }
    }

    @Test
    public void test_getAppUserList() {
        String appId = EnvHelper.getEnvValue("TEST_APP_ID");
        String code = EnvHelper.getEnvValue("TEST_CODE");
        try {
            ArrayList<Map<String, String>> response = OauthApiHelper.getAppUserList(appId, code, null);
            assertEquals(response.getClass(), ArrayList.class);
            assertEquals(response.get(0).getClass(), LinkedTreeMap.class);
            assertEquals(response.get(0).get("id").getClass(), String.class);
            assertEquals(response.get(0).get("open_id").getClass(), String.class);
        } catch (Exception e) {
            assertTrue(false);
        }
    }

    @Test
    public void test_getUserAppAccess() {
        String appId = EnvHelper.getEnvValue("TEST_APP_ID");
        String openId = EnvHelper.getEnvValue("TEST_OPEN_ID");
        String code = EnvHelper.getEnvValue("TEST_CODE");
        try {
            ArrayList<String> response = OauthApiHelper.getUserAppAccess(appId, openId, code, null);
            assertEquals(response.get(0).getClass(), String.class);
        } catch (Exception e) {
            assertTrue(false);
        }
    }

    @Test
    public void test_getUserIsAppAdded() {
        String appId = EnvHelper.getEnvValue("TEST_APP_ID");
        String openId = EnvHelper.getEnvValue("TEST_OPEN_ID");
        String code = EnvHelper.getEnvValue("TEST_CODE");
        try {
            Boolean response = OauthApiHelper.getUserIsAppAdded(appId, openId, code, null);
            assertEquals(response.getClass(), Boolean.class);
        } catch (Exception e) {
            assertTrue(false);
        }
    }

    @Test
    @SuppressWarnings("unchecked")
    public void test_getUserBaseInfo() {
        String appId = EnvHelper.getEnvValue("TEST_APP_ID");
        String openId = EnvHelper.getEnvValue("TEST_OPEN_ID");
        String code = EnvHelper.getEnvValue("TEST_CODE");
        try {
            Object response = OauthApiHelper.getUserBaseInfo(appId, openId, code, null);
            assertEquals(response.getClass(), LinkedTreeMap.class);
            Map<String, String> responseMap = (Map<String, String>) response;
            assertEquals(responseMap.get("nick").getClass(), String.class);
            assertEquals(responseMap.get("avatar").getClass(), String.class);
            assertEquals(responseMap.get("sex").getClass(), String.class);
        } catch (Exception e) {
            assertTrue(false);
        }
    }
}

File: php/tests/OauthApiHelperTest.php
<?php

namespace tests;

use PHPUnit\Framework\TestCase;
use tests\DotEnv;
use OauthApiHelper;

(new DotEnv(__DIR__ . '/.env'))->load();

final class OauthApiHelperTest extends TestCase
{
    public function test_post_proxy(): void
    {
        $getIpApi = "https://api-service.yuanzhibang.com/api/v1/Ip/getClientIp";
        $proxyAddress = getenv('TEST_PROXY_ADDRESS');
        $proxyAuth = getenv('TEST_PROXY_AUTH');
        $proxyIp = getenv('TEST_PROXY_IP');
        $proxy = ['ADDRESS' => $proxyAddress, 'AUTH' => $proxyAuth];
        $response = OauthApiHelper::post($getIpApi, [], [], $proxy);
        $this->assertSame($response, $proxyIp);
    }

    public function test_getApiByPath()
    {
        $api = OauthApiHelper::getApiByPath("/xxx");
        $expectApi = "https://oauth.yuanzhibang.com/xxx";
        $this->assertSame($api, $expectApi);
    }

    public function test_checkCode_valid()
    {
        $appId = getenv("TEST_APP_ID");
        $code = getenv("TEST_CODE");
        $response = OauthApiHelper::checkCode($appId, $code, "access_token");
        $this->assertArrayHasKey('expires_in', $response);
    }

    public function test_checkCode_invalid()
    {
        $appId = getenv("TEST_APP_ID");
        $code = "wrong code";
        $response = OauthApiHelper::checkCode($appId, $code, "access_token");
        $this->assertFalse($response);
    }

    public function test_getAppUserCount()
    {
        $appId = getenv("TEST_APP_ID");
        $code = getenv("TEST_CODE");
        $response = OauthApiHelper::getAppUserCount($appId, $code);
        $this->assertIsInt($response);
    }

    public function test_getAppUserList()
    {
        $appId = getenv("TEST_APP_ID");
        $code = getenv("TEST_CODE");
        $response = OauthApiHelper::getAppUserList($appId, $code);
        $this->assertIsArray($response);
        $this->assertArrayHasKey("id", $response[0]);
        $this->assertArrayHasKey("open_id", $response[0]);
    }

    public function test_getUserAppAccess()
    {
        $appId = getenv("TEST_APP_ID");
        $code = getenv("TEST_CODE");
        $openId = getenv("TEST_OPEN_ID");
        $response = OauthApiHelper::getUserAppAccess($appId, $openId, $code);
        $this->assertIsArray($response);
        $this->assertIsString($response[0]);
    }

    public function test_getUserIsAppAdded()
    {
        $appId = getenv("TEST_APP_ID");
        $code = getenv("TEST_CODE");
        $openId = getenv("TEST_OPEN_ID");
        $response = OauthApiHelper::getUserIsAppAdded($appId, $openId, $code);
        $this->assertIsBool($response);
    }

    public function test_getUserBaseInfo()
    {
        $appId = getenv("TEST_APP_ID");
        $code = getenv("TEST_CODE");
        $openId = getenv("TEST_OPEN_ID");
        $response = OauthApiHelper::getUserBaseInfo($appId, $openId, $code);
        $this->assertIsString($response['sex']);
        $this->assertIsString($response['avatar']);
        $this->assertIsString($response['nick']);
    }
}
